dvc/repo/datasets.py

```diff
@@ -0,0 +1,350 @@
+    return datetime.fromisoformat(d) if isinstance(d, str) else d
+
+
+def ensure(cls):
+    def inner(v):
+        return cls.from_dict(v) if isinstance(v, dict) else v
+
+    return inner
+
+
+class SerDe:
+    def to_dict(self: AttrsInstance) -> dict[str, Any]:
+        def filter_defaults(attr: Attribute, v: Any):
+            if attr.metadata.get("exclude_falsy", False) and not v:
+                return False
+            return attr.default != v
+
+        def value_serializer(_inst, _field, v):
+            return v.isoformat() if isinstance(v, datetime) else v
+
+        return asdict(self, filter=filter_defaults, value_serializer=value_serializer)
+
+    @classmethod
+    def from_dict(cls: type["Self"], d: dict[str, Any]) -> "Self":
+        _fields = fields(cast("type[AttrsInstance]", cls))
+        kwargs = {f.name: d[f.name] for f in _fields if f.name in d}
+        return cls(**kwargs)
+
+
+@frozen(kw_only=True)
+class DVCDataset(SerDe):
+    name: str
+    url: str
+    type: str
+    path: str = field(default="", converter=default_str)
+    rev: Optional[str] = None
+
+    def lock(self, repo, **kwargs):
+        from dvc.dependency import RepoDependency
+
+        def_repo = {
+            RepoDependency.PARAM_REV: self.rev,
+            RepoDependency.PARAM_URL: self.url,
+        }
+        dep = RepoDependency(def_repo, None, self.path, repo=repo)  # type: ignore[arg-type]
+        dep.save()
+        d = dep.dumpd()
+
+        repo_info = d[RepoDependency.PARAM_REPO]
+        assert isinstance(repo_info, dict)
+        rev_lock = repo_info[RepoDependency.PARAM_REV_LOCK]
+        return DVCDatasetLock(
+            name=self.name,
+            type=self.type,
+            url=self.url,
+            path=self.path,
+            rev=self.rev,
+            rev_lock=rev_lock,
+        )
+
+
+@frozen(kw_only=True)
+class DVCDatasetLock(DVCDataset):
+    rev_lock: str
+
+    def to_spec(self) -> "DVCDataset":
+        return DVCDataset.from_dict(self.to_dict())
+
+
+@frozen(kw_only=True)
+class DVCXDataset(SerDe):
+    name: str
+    url: str
+    type: str
+
+    @property
+    def name_version(self) -> tuple[str, Optional[int]]:
+        url = urlparse(self.url)
+        parts = url.netloc.split("@v")
+        assert parts
+
+        name = parts[0]
+        version = int(parts[1]) if len(parts) > 1 else None
+        return name, version
+
+    def lock(
+        self,
+        repo,  # noqa: ARG002
+        record: Optional["DatasetRecord"] = None,
+        version: Optional[int] = None,
+        **kwargs,
+    ):
+        if not record:
+            try:
+                from dvcx.catalog import get_catalog  # type: ignore[import]
+
+            except ImportError as exc:
+                raise DvcException("dvcx is not installed") from exc
+
+            name, version = self.name_version
+            catalog = get_catalog()
+            record = catalog.get_remote_dataset(name)
```

Comment:
`dvcx` datasets only work with SaaS at the moment. But, we can support it in dvc easily.
But supporting it in `Studio` is a different matter. So I have left it out.

There is no unique identifier for a dataset. There is a `created_at` timestamp. But this is only reliable for a remote dataset. Local dataset will have `created_at` timestamp from when it was `pull`-ed to the machine.

There is no provenance information for the dataset, either. The dataset could have come from a local dvcx instances, or a self-hosted Studio instances.
(I haven't looked at shadow datasets or made any effort to understand what they are.)

I'd really prefer a UUID or anything equivalent that is backed by some randomness rather than a timestamp.

We have another problem here. The datasets are local to each team in `dvcx`, but "Project" in Studio aren't. There could be projects in your private namespace, and some projects might be in a team (or, multiple teams') namespace. And you can have different dataset of the same name among different teams. Considering these federated/distributed hosts, and multiple teams, I'd not trust the timestamp at all. 

---

That said, I think it's best to ignore these things, unless dvcx starts taking these things seriously. :)

Reply:
I think it only makes sense for a remote/SaaS dataset. What's the point of versioning a local instance that could differ between users? If you want to pull a dataset locally, you can start with the remote dataset and track it with `dvc ds add`, then make that a dependency of a stage that uses the dvcx api to pull it locally.

Reply:
I guess my point is that there is a mismatch between how dvc and dvcx works. DVC will look at the SaaS for the dataset, but dvcx works based on a local index. 

So, a user might have a dataset `dogs` that was locally-created, but dvc will try to look into SaaS for it. The `DatasetQuery` runs locally based on that local dataset.
So we are not versioning what might actually get used.

Also, AFAIK the plan is to release CLI first and make some noise with it. So the user won't be able to use this integration unless we release SaaS at a later date (maybe end of this year?).

It'd be nice if we could always track the dataset that is available locally (and only fallback to SaaS if it does not exist), and have provenance information of where the dataset originated from so that we could link it back to the SaaS.

Reply:
So the example in https://github.com/iterative/dvc/pull/10316 doesn't make much sense as is since it will query the local instance, right?

There are a couple ways the workflow could make sense:
1. API methods to read from SaaS.
2. API to clone the dataset from SaaS to local.